### PR TITLE
Test: add testing for mqtt client client

### PR DESCRIPTION
**Pull request by andrewqiamzn**
### Why?
The DAB-ADB bridge currently has no tests. We want to add unit tests with Jest and achieve minimum 80% coverage.
### DOD
- [x] Achieve 80% test coverage `mqtt_client/client.ts`
### Test Plan
Screenshot 2022-12-22 at 10 02 29

### CODE REVIEW

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    });
+
+    describe("publish", () => {
+        it("should publish to a topic without a message and with no options", async () => {
+            const expectedPublishResponse = { cmd: "mockCmd", messageId: 1234, qos: 2 };
+
+            asMock(mockAsyncMqttClient.publish).mockResolvedValue(expectedPublishResponse);
+
+            asyncMqttClientOnCallBacks["connect"]();
+            const publishResponse = await (await testClient).publish(topics.DAB_MESSAGES);
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:64]: had to `await testClient` in every test as `testClient` wasn't awaited in the `before each`. If there is a cleaner way, please let me know!

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            const mockHandler = jest.fn(_message => Promise.resolve({ status: 200 }));
+            const inputTopic = topics.INPUT_KEY_PRESS_TOPIC;
+
+            asyncMqttClientOnCallBacks["connect"]();
+
+            await (
+                await testClient
+            ).handle(inputTopic, async (message: unknown) => {
+                return await mockHandler(message);
+            });
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:236]: A shorter syntax would be `.handle(inputTopic, mockHandler)`

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            await (await testClient).end();
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:237]: A way to test that the handlers have been removed is to send a message after `end` is called and verify that it's not received by the handler. Something like:
```
asyncMqttClientOnCallBacks["message"](inputTopic, "a message");
expect (mockHandler).haveBeenCalledWith("a message");
await testClient.end();
mockHandler.mockClear();
asyncMqttClientOnCallBacks["message"](inputTopic, "another message");
expect (mockHandler).not.haveBeenCalled();
```

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            await requestResponse;
+            await (await testClient).end();
+
+            expect(mockAsyncMqttClient.subscribe).toBeCalledWith(`${inputTopic}/+`);
+            expect(mockAsyncMqttClient.publish).toBeCalledWith(
+                `_response/${inputTopic}`,
+                JSON.stringify({ status: 200 }),
+                { qos: 2, retain: false }
+            );
+            expect(mockHandler).toBeCalled();
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:201]: Nit: better to check that the handler has been called with the right params

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            await asyncMqttClientOnCallBacks["message"](`${inputTopic}/*`, bufferObject, {
+                qos: 0,
+                cmd: "publish",
+                dup: false,
+                retain: false,
+                payload: Buffer.from('{"status":200,"message":{"keyCode":"KEY_DOWN"}}', "utf8"),
+                topic: inputTopic
+            });
+
+            await requestResponse;
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:192]: Add an assertion on the response to validate it is the expected one

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+            expect(mockAsyncMqttClient.publish).toBeCalledWith(
+                `_response/${inputTopic}`,
+                JSON.stringify({ status: 200 }),
+                { qos: 2, retain: false }
+            );
+            expect(mockHandler).toBeCalled();
+            expect(mockAsyncMqttClient.end).toBeCalled();
+        });
+
+        it("should not publish if the handler failed to receive a request topic", async () => {
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:205]: This test is a bit unclear. My understanding is that it should be "should not publish a response if there is no handler associated to the topic". Is it right? In which case it's better to set an explicit topic in the message `topic: "someOtherTopic"` so that it's clearer

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        testClient = connect(mockUri);
+    });
+
+    describe("publish", () => {
+        it("should publish to a topic without a message and with no options", async () => {
+            const expectedPublishResponse = { cmd: "mockCmd", messageId: 1234, qos: 2 };
+
+            asMock(mockAsyncMqttClient.publish).mockResolvedValue(expectedPublishResponse);
+
+            asyncMqttClientOnCallBacks["connect"]();
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:55]: This is replicated in all tests, because all APIs (except `connect()`) need to be called after `connect()` has been called. Consider to add a `describe("after connect")` block and put this statement in the `beforeEach`.  You can leave only the tests of `connect()` outside this describe block. 

Also, we would need a test for the case when these APIs are called before calling `connect()`. I imagine they will fail? throw an error? These tests could go in another describe block "before connect"

**Review by giotar — CHANGES_REQUESTED** [review]

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 1 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:56]: No, if I do `testClient = await connect(mockUri)`, it would never resolve as the i would never call the the `on` call back here. This is why I couldn't await it. 
image

Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:56]: I also tried doing this, but because i get a promise, typescript raises an error.
image

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+import * as topics from "../../dab/dab_topics";
+import mqtt, { OnErrorCallback, OnMessageCallback } from "async-mqtt";
+import { asMock } from "../../../__tests__/TestUtils";
+import { DabKey } from "../../adb/adb_keymap";
+import { v4 as uuidv4 } from "uuid";
+
+type EventCallback = (topic?: string, message?: string) => void;
+
+interface AsyncMqttClientOnCallBacks {
+    connect: EventCallback;
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:11]: Good point, fixed in the next commit.

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+                return await mockHandler(message);
+            });
+            await (await testClient).end();
+
+            expect(mockAsyncMqttClient.unsubscribe).toBeCalledWith(`${inputTopic}/+`);
+            expect(mockAsyncMqttClient.end).toBeCalled();
+        });
+    });
+
+    describe("connect", () => {
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:244]: Maybe this wasn't clear, this is testing the exported function `connect` that creates and returns a `Client` object (screenshot below). The actual client object doesn't have any `connect` methods.

image

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 2 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:236]: Ah nice, fixed in the next commit!

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+            asyncMqttClientOnCallBacks["error"](new Error("this is an error"));
+
+            await testClient;
+
+            expect(mockAsyncMqttClient.removeListener).toBeCalled();
+        });
+
+        it("should remove the error listener and call end if there is an error event and if it is not connected and not initialised", async () => {
+            asyncMqttClientOnCallBacks["error"](new Error("this is an error"));
+            asyncMqttClientOnCallBacks["connect"]();
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:256]: Thanks for spotting this. I don't need it and have updated the test to check for an error instead.

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 3 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:237]: Got it! We're basically checking the before and after effects of removing the listener! Thanks

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 5 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:192]: This is just a `Promise<void>`, so it might not be that useful to assert this. I'll update the variable name to `requestPromise` to clarify that it needed to be awaited.

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 6 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:205]: This test is testing the code flow for when there is no request topic (screenshot). I can make it clearer by extracting it to a variable `emptyRequestTopic = ""`

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 7 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:55]: I've updated accordingly for the first point, thanks! For the second point, the only way the function `connect` throws an error is if the `onError` callback is called, and subsequently rejects with an error.

image

Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:55]: I've updated my test to check for this error here

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 1 above)
Comment (giotar) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts]: You could probably do 
```
const connectPromise = connect(mockuri);
asyncMqttclientOnCallbacks["connect"]();
testclient = await connectPromise;
```

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 7 above)
Comment (giotar) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:49]: Yes, that's correct. 

As I have written in another comment, I would only add to this beforeEach 
```
testClient = await connectPromise
```
so to not have it in each test

Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:49]: Neat! I've updated :)

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+                expect(publishResponse).toEqual(expectedPublishResponse);
+            });
+
+            it("should publish to a topic with a message and with options", async () => {
+                const mockMsg = { message: "mockMessage" };
+                const expectedPublishResponse = { cmd: "mockCmd", messageId: 1234, qos: 2 };
+
+                asMock(mockAsyncMqttClient.publish).mockResolvedValue(expectedPublishResponse);
+                const publishResponse = await (
+                    await testClient
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:73]: Leftover `await`?

**Review by giotar — APPROVED** [review]

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts (hunk 11 above)
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:73]: Ah! Thanks for spotting :)

File: adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+import { DabKey } from "../../adb/adb_keymap";
+import { v4 as uuidv4 } from "uuid";
+
+interface AsyncMqttClientOnCallBacks {
+    connect: () => void;
+    error: OnErrorCallback;
+    message: OnMessageCallback;
+}
+
+jest.setTimeout(30000);
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:14]: About 5.5 seconds

Looks like one of the tests (line 157) takes 5 seconds which is the one using the fake timer.


Screenshot 2023-01-11 at 16 09 45

[the remark this answers is not in the document]
Comment (amzn-stenlaan) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:14]: I think it's using real timers as the MQTT timeout is also supposed to be 5 seconds. You can test that by increasing timeout to 30 seconds and seeing what happens.

[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/mqtt_client/__tests__/client.test.ts:14]: Yeah you're right. It does increase to 30 seconds. I tried using `jest.runAllTimers()` [here to try to counteract this](https://github.com/device-automation-bus/dab-tools/pull/14/files/77bc0141bff9d637fdca4ef6b08ea36bd760d701#diff-f7261cd1b3f876661c3ccb98c70ee595b2da761435a0130e3f6bf38bae145eafR161)

File: adapters/dab-adb-bridge/src/lib/mqtt_client/client.ts [hunk 13]
```diff
@@ -28,7 +28,6 @@ import { EventEmitter2, ListenerFn } from "eventemitter2";
 import { IClientPublishOptions } from "mqtt";
 import { IPublishPacket } from "mqtt-packet";
 import { HandlerFunction, HandlerSubscription } from "./index";
-import { setTimeout, clearTimeout } from "timers";
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/mqtt_client/client.ts:31]: The problem was that nodejs timers were imported, which jest couldn't mock, hence jest's fake timers didn't work.

**Review by amzn-stenlaan — APPROVED** [review]
